### ISSUE TITLE
Multiple types of Array?

### ISSUE THREAD

**Issue by chasebish** [rucha2398/cs4500-a1-p2#5]
According to this line in the assignment:
`there should be Array classes which supports int, float, bool and String`
would require a few changes to the spec since it does not currently support int/float/bool. How would you want to go about changing it?

A few ideas:
- create an enum and pass it through with the constructor
- arrays can take any value when they are empty, but restricted once they have >= 1 elements to whatever is first in the list
- create wrapper classes around Array, one for each type

These all seem messy, so the best implementation may be to store everything using a linked list and then not restrict based on datatype? This [Piazza](https://piazza.com/class/k51bluky59n2jr?cid=320) post briefly discusses this problem but it's not very helpful

**Comment by chasebish** [comment]
Additionally, if you end up implementing String, I previously linked the wrong file
[string.h](https://github.com/chasebish/cs4500_assignment1_part2/blob/master/string.h)

**Comment by chasebish** [comment]
We just spoke with Jan, and it seemed like he was eluding towards having 4 separate APIs, IntArray, FloatArray, BoolArray, and StringArray. We are working on implementing these for our own spec since that is what we did for Assignment1, so we can share it with you when we are done.

**Comment by chasebish** [comment]
After thinking about it and discussing it further, we found it's probably best to instead implement our own separate classes, Integer, Float, Boolean, and String which all extend Object. This will only require limited changes to the API

**Comment by chasebish** [comment]
However, it might be important for you to have your own APIs for Integer, Float, Boolean, and String, that way the test file can be updated accordingly

**Comment by jjones-99** [comment]
We agree with @chasebish on implementing wrapper class APIs for `Integer`, `Float`, `Boolean`, and `String`. The test file should be updated to cover cases of using Arrays with these types, and to that end, each type should have an API spec defined by your team.

**Comment by zcorning** [comment]
Right now, we are operating under the assumption that people can use wrapper classes for adding different types to the array. If this changes in the future or it becomes clear that the array should accept primitive types, we will change the API, but for now, it seems unnecessary.

Since the array only needs to include the object.h class for the API, the tests should reflect that functionality. If you want to test the wrapper classes specifically, feel free to do so since that is part of your own implementation. However, the array should operate the same over all object types, so the array tests as part of the API are only over objects.

In all, we think this repo should contain just the APIs necessary for interfacing with an array. All other implementation of types, etc. should be specific to your implementation.

**Comment by chasebish** [comment]
I disagree that types should be specific to implementation. Since the spec requires that an Array must work with bool, float, int, and String, those should all be tested and included with the API.

**Comment by jjones-99** [comment]
Hey @zcorning, while we agree that this—not needing separate Array types for each type of value—is a valid design choice, we will draw attention to Jan's response to [Piazza @331](https://piazza.com/class/k51bluky59n2jr?cid=331), where he says "...imagine that you need to store number 42 into newly created array of integers. What is the most natural way to write the two lines that create the array and push_back the number?"

While this specification works reasonably cleanly when you push values (assuming us implementation pairs create wrapper classes):
```
Array* arr = new Array();
arr->push_back(new Integer(7));
```

It does bring up the more 'gross' requirement of casting in some manner whenever you retrieve values:
```
Integer* i = dynamic_cast<Integer*>(arr->get(0));

// It would be easier to not have to cast repeatedly.
Integer* i = arr->get(0);  //Doesn't work.
```

_Also, in a later followup post, Jan mentions, "But, yes, I do believe that from an API it is clearly better to avoid the cast (because it is unchecked and you could get it wrong)."_

That's a design choice that we'll leave to you, we just wanted to bring it to your attention.

### The More Important Piece

That being said, however, the spec team is responsible for writing tests for all use cases they foresee as being important (see Instructor Note in [Piazza @320](https://piazza.com/class/k51bluky59n2jr?cid=320)). Because the assignment description includes the ability to store ints, floats, bools, and Strings in an Array (four use cases), there should be tests written for those four types.

In writing those tests, you will have to make the decision on how the API functions. For instance, with the current API, it's actually impossible to write a test using an Array that stores ints, floats, and bools.

### TL;DR (Sorry the above is so long 😭)

- The current API itself kind of fulfills the requirements, but from the user's point of view is a little cumbersome.
- Regardless of that, the test cases should cover the use cases of Arrays of ints, bools, and floats—which is currently impossible with this API.
- Thus, there should be some specification of how one would use an Array in those cases. (e.g., `IntArray` and similar types without casting, `Array` and casting with `Integer` and similar wrappers, etc.)

**Comment by chasebish** [comment]
I agree with @jjones-99

I think we are both on the same page that the API is a bit incomplete. At this point I see two possibilities
1. Implementing 4 different arrays
2. Implementing 4 separate classes (Bool, Float, Integer, String)

The reason one of these two is necessary because the tests will be incomplete without them. Just pushing Objects onto Arrays isn't sufficient because that would allow for different types to be pushed onto the same array.

I think both of us would appreciate if you guys could choose one rather soon because I would like to have a few days for implementation. Sorry for the continuous posting.

**Comment by rucha2398** [comment]
We are currently working on adding subclasses for FloatArray, IntArray, BoolArray and StringArray. Will push code soon, we are just finishing up writing the tests! Thank you for your patience!!! Hope this didn't slow you down too much!!

**Comment by rucha2398** [comment]
Sorry for the late push, it should all be there now! Let us know if this works.

**Comment by rucha2398** [comment]
If any of the tests are wrong, please let us know ASAP so we can fix them for you!!

**Comment by jjones-99** [comment]
This looks all good to us! There's one tiny issue with the tests, but we'll open it in a new issue to not overwhelm this issue.
